src/test/java/org/snomed/snowstorm/rest/CodeSystemControllerTest.java

```diff
@@ -0,0 +1,115 @@
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.LinkedHashMap;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+
+
+@ActiveProfiles("test")
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT, classes = TestConfig.class)
+class CodeSystemControllerTest extends AbstractTest {
+    @LocalServerPort
+    private int port;
+
+    @Autowired
+    private TestRestTemplate restTemplate;
+
+    @Autowired
+    private CodeSystemService codeSystemService;
+
+    @Autowired
+    private BranchService branchService;
+
+    @Test
+    public void listCodeSystems_ShouldReturnCodeSystem_WhenCodeSystemExists() {
+        //given
+        String branchPath = "MAIN/ProjectA";
+        String shortName = "SNOMEDCT-PROJECT-A";
+        String requestUrl = listCodeSystems(branchPath);
+        givenCodeSystemExists(shortName, branchPath);
+        givenCodeSystemVersionExists(shortName, 20200731, "2020 July.");
+
+        //when
+        ItemsPage<?> result = this.restTemplate.getForObject(requestUrl, ItemsPage.class);
+
+        //then
+        assertNotNull(result);
+        assertEquals(1, result.getTotal());
+    }
+
+    @Test
+    public void listCodeSystems_ShouldShowLatestVersionDependsOnPreviousVersion() {
+        //given
+        String mainShortName = "MAIN";
+        String projectPath = "MAIN/ProjectA";
+        String projectTaskPath = "MAIN/ProjectA/TaskA";
+        String projectShortName = "SNOMEDCT-TEST-SHORTNAME";
+        String requestUrl = listCodeSystems(projectTaskPath);
+
+        givenBranchExists(projectPath);
+        givenBranchExists(projectTaskPath);
+
+        givenCodeSystemExists(mainShortName, mainShortName);
+        givenCodeSystemExists(projectShortName, projectPath);
+
+        givenCodeSystemVersionExists(mainShortName, 20200131, "2020 January.");
+        givenCodeSystemVersionExists(mainShortName, 20210131, "2021 January.");
+        givenCodeSystemVersionExists(projectShortName, 20190731, "2019 July.");
+        givenCodeSystemVersionExists(projectShortName, 20200131, "2020 January.");
+        givenCodeSystemVersionExists(projectShortName, 20200731, "2020 July.");
+
+        //when
+        ItemsPage<ItemsPage<?>> result = this.restTemplate.getForObject(requestUrl, ItemsPage.class);
+        Integer dependantVersionEffectiveTime = getDependantVersionEffectiveTime(result);
+
+        //then
+        assertEquals(1, result.getTotal());
+        assertEquals(20200131, dependantVersionEffectiveTime);
+    }
+
+    private String listCodeSystems(String branchPath) {
+        return "http://localhost:" + port + "/codesystems/?forBranch=" + branchPath;
+    }
+
+    private Integer getDependantVersionEffectiveTime(ItemsPage<ItemsPage<?>> result) {
+        Collection<ItemsPage<?>> items = result.getItems();
+
+        if (items instanceof ArrayList) {
```

Comment:
If you look at this if-block in your IDE, you'll see a lot of warnings about generics and unchecked assignments. I couldn't work out the correct way to remove them and I could only shift the problem and not resolve it. If you wouldn't mind, it would be great seeing your solution.

Reply:
I've provided two solutions here.. one low level and one using domain objects..
On line 79..
```
        ItemsPage<?> result = this.restTemplate.getForObject(requestUrl, ItemsPage.class);
```
Then 
```
    private Integer getDependantVersionEffectiveTime(ItemsPage<?> result) {
    	@SuppressWarnings("unchecked")
		List<Map<String, Object>> items = (List<Map<String, Object>>) result.getItems();
		Map<String, Object> item = items.get(0);
		@SuppressWarnings("unchecked")
		Map<String, Object> latestVersion = (Map<String, Object>) item.get("latestVersion");
		return (Integer) latestVersion.get("dependantVersionEffectiveTime");
    }
```
Using `@SuppressWarnings` is a bit of a workaround for when you really don't know the java type that you're dealing with up front, or in this case where the Map contains values with a mixture of classes, and need to cast.

The other approach is to use a ParameterizedTypeReference to get the response mapped straight into the domain objects like this:
```
ResponseEntity<ItemsPage<CodeSystem>> result = this.restTemplate.exchange(requestUrl, HttpMethod.GET, null, new ParameterizedTypeReference<>() {});
```
I haven't tested either of these examples but you should get the idea.